libraries/chain/controller.cpp

```diff
@@ -1226,36 +1374,37 @@ struct controller_impl {
                throw *trace->except;
             }
 
-            EOS_ASSERT( pending->_pending_block_state->block->transactions.size() > 0,
+            EOS_ASSERT( trx_receipts.size() > 0,
                         block_validate_exception, "expected a receipt",
                         ("block", *b)("expected_receipt", receipt)
                       );
-            EOS_ASSERT( pending->_pending_block_state->block->transactions.size() == num_pending_receipts + 1,
+            EOS_ASSERT( trx_receipts.size() == num_pending_receipts + 1,
                         block_validate_exception, "expected receipt was not added",
                         ("block", *b)("expected_receipt", receipt)
                       );
-            const transaction_receipt_header& r = pending->_pending_block_state->block->transactions.back();
+            const transaction_receipt_header& r = trx_receipts.back();
             EOS_ASSERT( r == static_cast<const transaction_receipt_header&>(receipt),
                         block_validate_exception, "receipt does not match",
-                        ("producer_receipt", receipt)("validator_receipt", pending->_pending_block_state->block->transactions.back()) );
+                        ("producer_receipt", receipt)("validator_receipt", trx_receipts.back()) );
          }
 
-         finalize_block();
+         auto block_ptr = finalize_block();
 
          // this implicitly asserts that all header fields (less the signature) are identical
-         EOS_ASSERT(producer_block_id == pending->_pending_block_state->header.id(),
-                   block_validate_exception, "Block ID does not match",
-                   ("producer_block_id",producer_block_id)("validator_block_id",pending->_pending_block_state->header.id()));
+         auto id = block_ptr->id();
+         EOS_ASSERT( producer_block_id == id, block_validate_exception, "Block ID does not match",
+                     ("producer_block_id",producer_block_id)("validator_block_id",id) );
 
-         // We need to fill out the pending block state's block because that gets serialized in the reversible block log
-         // in the future we can optimize this by serializing the original and not the copy
+         auto&& ab = pending->_block_stage.get<assembled_block>();
 
```

Comment:
Here and in `controller::finalize_block` I don't like the inconsistency of returning signed_block from finalize_block but also reaching into `pending->_block_stage.get<assembled_block>()`. Seems you could either not return anything from `finalize_block` or return the block_stage. In either case if id was added to assembled_block then it would not have to be recalculated via `id()`.